chat.go

```diff
@@ -0,0 +1,190 @@
+// Package chat implements a chat server & client
+package chat
+
+import (
+	"bufio"
+	"fmt"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+)
+
+// Server represents a chat server
+type Server struct {
+	mutex         sync.Mutex
+	listener      net.Listener
+	running       bool
+	host          string
+	port          int
+	Logger        logger // with standard logger can be extended with logrus
+	ListenAddress string
+}
+
+// Logger enables a customization of the log function
+type logger interface {
+	Println(v ...interface{})
+}
+
+// StandardLogger defines a standard logger that implements the Logger interface
+type StandardLogger struct {
+	timeFormat string
+}
+
+// Client represents a chat client
+type Client struct {
+	connection net.Conn
+}
+
+// StartServer returns a pointer to a running server on localhost and random port
+func StartServer(options ...func(*Server)) (*Server, error) {
```

Comment:
Yes sure. But I have a question though.
To implement option 1. should I move the listener creation and the run goroutine into the listen and serve ?
Otherwise, it will be difficult for the user to customize the port as the StartServer immediatly starts the server.